### PR TITLE
fix(events): only trigger change when there is an actual change

### PR DESCRIPTION
**Pull request by bobylito**
In order to fix this issue, we needed to actually fix two things:
 - make sure that the mechanics in the Helper, check if there is a `change` and not trigger all the time. This implies a refactoring on the Helper side of all the mutating methods and make sure that before triggering the change event, we check the changes.
 - make sure that we are able to determine that there has been a change. This implies to verify that all updating methods on the searchParameters are not returning new instance of the state if this is not necessary.

The first point was pretty easy to implement. For the second, it was trickier. The clears were not scarce at all.

Also please the implementation of the _clearRefinements and clear (in refinementList), the usage with function will create a new structure but will not return if it yields the same value. This is a bit of trick but couldn't find a better solution for that. If you have another approach, I'd be happy to learn about it :)

### CODE REVIEW

File: src/SearchParameters/index.js [hunk 1]
```diff
@@ -620,13 +620,21 @@ SearchParameters.prototype = {
...
-    return this.setQueryParameters({
+    var patch = {
       numericRefinements: this._clearNumericRefinements(attribute),
       facetsRefinements: clear(this.facetsRefinements, attribute, 'conjunctiveFacet'),
       facetsExcludes: clear(this.facetsExcludes, attribute, 'exclude'),
       disjunctiveFacetsRefinements: clear(this.disjunctiveFacetsRefinements, attribute, 'disjunctiveFacet'),
       hierarchicalFacetsRefinements: clear(this.hierarchicalFacetsRefinements, attribute, 'hierarchicalFacet')
-    });
+    };
+    if (patch.numericRefinements === this.numericRefinements &&
```
Comment (bobylito) [inline src/SearchParameters/index.js]: I probably could move this logic to the `setQueryParameters`

File: src/SearchParameters/RefinementList.js [hunk 2]
```diff
@@ -96,19 +96,30 @@ var lib = {
...
+      var hasChanged = false;
+
+      var newRefinementList = reduce(refinementList, function(memo, values, key) {
         var facetList = filter(values, function(value) {
           return !attribute(value, key, refinementType);
         });
 
-        if (!isEmpty(facetList)) memo[key] = facetList;
+        if (!isEmpty(facetList)) {
+          if (facetList.length !== values.length) hasChanged = hasChanged || true;
```
[the remark this answers is not in the document]
Comment (bobylito) [reply src/SearchParameters/RefinementList.js]: `true`. That's a good simplification.

File: src/SearchParameters/index.js [hunk 3]
```diff
@@ -1676,6 +1694,10 @@ SearchParameters.prototype = {
     );
     var path = refinement.split(separator);
     return map(path, trim);
+  },
+
+  toString: function() {
+    return JSON.stringify(this, null, 2);
```
[the remark this answers is not in the document]
Comment (vvo) [reply src/SearchParameters/index.js]: Any example of such debug session? Again curious that's it

[the remark this answers is not in the document]
Comment (iam4x) [reply src/SearchParameters/index.js]: @vvo This is useful when you want to console.log all the things and in a React context I will do this:

```jsx
<pre>
  { JSON.stringify(this.props, null, 4) }
</pre>
```

File: src/SearchParameters/index.js (hunk 1 above)
Comment (iam4x) [reply src/SearchParameters/index.js]: I think it's better here, `clearRefinements` should hold this specific logic not the `setQueryParameters`.

**Comment by bobylito** [comment]
What is a design mistake @iam4x ? :)

**Comment by iam4x** [comment]
I mean that if we plan to re-work the helper from the start we should have these cases in the first design process. It will make things much easier.

But you are right, sometimes you can't just predict this kind of outcome 👍

**Comment by vvo** [comment]
> But you are right, sometimes you can't just predict this kind of outcome 👍

Oh yes getting the API right from the beginning is hard because you are never right at the right time (if you need me to write again `right` let me know, right.)

 This is a good iteration, API becomes better by iterating and also with good initial design, true

**Comment by rayrutjes** [comment]
Thanks a lot for this massive PR @bobylito !

**Comment by bobylito** [comment]
My pleasure @rayrutjes It always feels good to go back to this project :)

**Pull request merged**